### ISSUE TITLE
Binding to byte array?

### ISSUE THREAD

**Issue by Dunge** [XamlAnimatedGif/XamlAnimatedGif#77]
I have a C# object containing a byte array property containing the gif data instead of loading gif files on disk (so I can work in-memory and transfers gifs over wcf, without ever saving them as files on disk).

I was previously using WpfAnimatedGif and it worked perfectly, I could just do a 
`gif:ImageBehavior.AnimatedSource="{Binding Data}"`
Where Data is my byte[] property.

Now I encountered the problem with memory leak when displaying more than a few gif at once, ended up on your blog and noticed I had to switch to XamlAnimatedGif to fix it. Binding a byte array to SourceUri of SourceStream doesn't work anymore.

I tried a simple method (not sure if it's safe, should be disposed?):
```
        // Raw gif file as byte array
        public byte[] Data { get; set; }
        public MemoryStream DataStream
        {
            get
            {
                return new MemoryStream(Data);
            }
        }
```

and

`gif:AnimationBehavior.SourceStream="{Binding DataStream}"`

Unfortunately, I don't see anything get displayed, and no error message either.

**Comment by thomaslevesque** [comment]
Hi @Dunge,

Are you sure that `Data` is never null when the `DataStream` property is called? Also, you can subscribe to the `Error` event to detect GIF loading and rendering errors.

**Comment by Dunge** [comment]
Thanks for the reply.

From a breakpoint, I can assure Data is never null when called.

I played a bit around by transforming DataStream to be a NotifyPropertyChanged property, and I can see that if I set the DataStream property on an event (button click, etc.) it works fine, but if it is set during the viewmodel class constructor before the window/control is shown, I get an error "Cannot access a closed Stream.":

   at System.IO.__Error.StreamIsClosed()
   at System.IO.BufferedStream..ctor(Stream stream, Int32 bufferSize)
   at System.IO.BufferedStream..ctor(Stream stream)
   at XamlAnimatedGif.Extensions.StreamExtensions.AsBuffered(Stream stream)
   at XamlAnimatedGif.AnimationBehavior.InitAnimation(Image image)

Worth saying that I never close the Stream, the XamlAnimatedGif library probably does at some point.

This is unfortunate, because I would like to put the animated gif in a DataTemplate inside a ListBox that the binded ItemsSource collection is a simple list of object containing byte arrays of gifs, without having to do code-behind to create events that would allocate streams and dispose them.

**Comment by thomaslevesque** [comment]
Could you show more of your code?

> Worth saying that I never close the Stream, the XamlAnimatedGif library probably does at some point.

I think it could happen if the binding happens more than once. The second time, the previous `Animator` is disposed, which in turn disposes the stream. The stream should probably not be closed if it was provided via `SourceStream`. I'll fix it soon, unless you want to take a stab at it.

**Comment by Dunge** [comment]
Sorry for the delay. I'm just writing to say that I did not forget about this ticket, but am currently busy on other projects.

I don't think the problem is a second binding, because it does that the first time the window display. When I change the binding to a new stream is when it work, not the other way around.

My code is in a pretty big project with tons of dependencies and I can't share it, so I decided to try to reproduce my problem in a new separate project, and everything works fine. So the problem might lie on my side. I'll try to find more time to trim it down until I can see what's causing it. To be continued.

**Comment by Dunge** [comment]
Salut Thomas,

I have progress on this issue and now believe it is indeed a problem in the library, but very far and much more deep than what I first expected.

I ended up loading the library source instead of the nuget package to debug what was happening. I was having an InvalidSignatureException getting catched in the InitAnimationAsync method. I found the thrower was in the GifHeader decoding validating that the first 3 bytes are 'GIF' even if my stream first 3 bytes really are 'GIF'. I then realized that it ran this code twice, both reading in parallel asynchronously from the same stream and one progress would impact the cursor position on the stream for the loading of the second run. I tracked back the source to the AnimationBehavior::InitAnimation() being called twice due to dependencyproperty changing, once for the SourceStream property, once for the RepeatBehavior property.

Removing the RepeatBehavior property made things work, and adding it on my external progress broke it. When I was re-affecting the SourceStream property on a button click, only one was refresh so the loading was working properly.

So in short, you read the same stream simultaneously because two property change triggers simultaneously, and it cause problems. Now, I know the cause, but I'm not sure how to fix it. You know your project better than me, would it be possible you look at it?

I created a VERY simple vs2013 solution reproducing it, you can download it here:
http://www.jamlogic.com/download/Gif.zip

Merci!

**Comment by thomaslevesque** [comment]
Ouch, that's pretty bad... thanks for the investigation, I'll see what I can do.

**Comment by thomaslevesque** [comment]
OK, I think I have a fix: https://github.com/XamlAnimatedGif/XamlAnimatedGif/pull/82

There were actually two different issues

- `Animator` was always disposing the source stream, which caused a problem in your "byte array" scenario: when the `RepeatBehavior` was changed, it caused the animation to be reinitialized, disposing the previous `Animator`, which caused the stream to be disposed as well. In fact, the library should *never* dispose a stream that was provided directly by the user (via the `SourceStream` property), because it doesn't own it. Now `Animator` disposes the stream only if it was created by the library from a URI.
- Setting the `RepeatBehavior` caused the animation to be reinitialized. It seemed easier to do it this way initially, but it was inefficient and caused the issue you described above. Now it no longer recreates the whole animation, it just signals the `Animator` to change its `RepeatBehavior`. This is more efficient and avoids the concurrent stream access.

I haven't merged the fix yet; could you try it and confirm it works for you?

**Comment by Dunge** [comment]
Thank your very much for the quick fix, it is appreciated.
With the new code, the gif now display properly.

But now, when I close the window hosting the gif control, I receive System.OperationCanceledException and System.Threading.Tasks.TaskCanceledException. I never really worked with Cancellation Token, so I'm not sure if this is normal or not. Should I send a cancellation signal in my window closing event to tell the lib to stop playing the gif?

Also, I don't want to ask too much, but I'm coming back to my original question. What do you think about the idea of directly binding to a byte array and let the lib handle the stream internally (like WpfAnimatedGif was doing, and the normal .NET Image control Source property support)? Having to handle the creation and dispose of the stream in an intermediary class between my data model (raw bytes that get serialized) and the view (user control) is giving me a bit of trouble. It's easy when you have a view model for the window showing a single image control but it becomes an headache in a complex project like mine. I have a list of objects of multiple types bound to a listbox. Each of my type have DataTemplate saying how to display it, and only when I get the animation type I need to display the gif control, but in this case it need to be a "view model gif" to handle the stream lifespan instead of my raw data so I need to transform them into a sub class to handle said stream, which is not quite easy. The "easy" way I made it (in my first post, always creating a new memorystream) leak and expose a property in my data that can't get serialized. I hope this make sense.

On a separate note, how fast do you update the nuget library when releasing a fix / new version? If I want to release something next week, do you recommend I include the project from source instead of the nuget package?

Thanks again, sorry for making things complicated :)

**Comment by thomaslevesque** [comment]
> But now, when I close the window hosting the gif control, I receive System.OperationCanceledException and System.Threading.Tasks.TaskCanceledException

Where are you seeing this? These exceptions can be thrown internally when the animation is stopped, but they should be surfaced outside the library. Are you sure they're not just first-chance exceptions that are caught later?

Anyway, I don't see these exceptions in the test project you sent.

> Also, I don't want to ask too much, but I'm coming back to my original question. What do you think about the idea of directly binding to a byte array and let the lib handle the stream internally (like WpfAnimatedGif was doing, and the normal .NET Image control Source property support)?

WpfAnimatedGif wasn't doing anything to handle this; I think it must be done automatically by WPF when you bind an `ImageSource` property to a byte array.

Honestly, I don't really like the idea. I worry that if I start adding too many features for specific use cases, the library will become bloated and hard to maintain. In this case, there's already a way to do what you want (using a `MemoryStream`)

> Having to handle the creation and dispose of the stream in an intermediary class between my data model (raw bytes that get serialized) and the view (user control) is giving me a bit of trouble. [...]

I haven't seen your code, but it looks like you're probably overthinking it... The lifetime of the stream isn't really an issue; streams should generally be disposed, but for `MemoryStream`s, it doesn't really matter, because they don't use any unmanaged resources, and are practically free (they just wrap the byte array passed to the constructor). They will be garbage collected after use anyway. So you could use a binding converter to wrap the byte array in a `MemoryStream`.

> On a separate note, how fast do you update the nuget library when releasing a fix / new version? If I want to release something next week, do you recommend I include the project from source instead of the nuget package?

Pretty fast. I typically make a new release every time I fix an issue. When I merge the branch, the new package can be on NuGet minutes later.

**Comment by Dunge** [comment]
Ok, thanks for the clarifications.

You are right the exceptions only appear as "first-chance" in the output window.
I also just read about MemoryStream disposing, and you are right they don't leak anything, so that's great. It seems a bit strange not to dispose something implementing IDisposable, but I like the idea of using a Converter for this, much easier than what I was going to do.

**Comment by thomaslevesque** [comment]
OK, I'll merge the fix then

**Comment by thomaslevesque** [comment]
Version 1.1.6 is deployed on NuGet (might be a few minutes before it appears in search results, because indexing isn't complete yet)
